src/typecheck/eq.rs

```diff
@@ -0,0 +1,356 @@
+//! `let Alias' = Foo "bar" "baz" in ...`, or `let Alias' = Alias in `.
+//!
+//! ## Recursion
+//!
+//! We must refrain from following all variables links blindly, as their could be cycles in the
+//! graph leading to an infinite loop:
+//!
+//! ```nickel
+//! {
+//!   Foo = Bar,
+//!   Bar = Foo,
+//! }
+//! ```
+//!
+//! Because we just follows variables, and don't apply functions, we can detect cycles while
+//! walking the graph. Still, as it is performed potentially many times during typechecking, type
+//! equality ought to stay reasonably cheap. We choose to just set an arbitrary limit (the gas) on
+//! the number of variable links that the type equality may follow. Doing so, we don't have to
+//! worry about loops anymore.
+//!
+//! ## Equality on terms
+//!
+//! The terms inside a type may be arbitrarily complex. Primops applications, `switch`, and the
+//! like are quite unlikely to appear inside an annotation (they surely appear inside contract
+//! definitions). We don't want to compare functions syntactically either. The spirit of this
+//! implementation is to equate aliases or simple constructs that may appear inlined inside an
+//! annotation (applications, records, primitive constants and arrays, mostly) in a structural way.
+//! We first test for physical equality (both as an optimization and to detect two variables
+//! pointing to the same contract definition in the AST). If the comparison fails, we do a simple
+//! structural recursion, unfolding simple forms and following variables with a limited number of
+//! times. For anything more complex, we return false.
+
+use super::*;
+
+/// The maximal number of variable links we want to unfold before abandoning the check. It should
+/// stay low, but has been fixed arbitrarily: feel fee to increase reasonably if it turns out
+/// legitimate type equalities between simple contracts are unduly rejected in practice.
+pub const MAX_GAS: u8 = 8;
+
+#[derive(PartialEq, Clone)]
+pub struct TermEnvironment<'a>(
+    pub crate::environment::Environment<Ident, (&'a RichTerm, TermEnvironment<'a>)>,
+);
+
+/// State threaded through the type equality computation.
+#[derive(Copy, Clone, Default)]
+struct State {
+    /// Used to generate temporary rigid type variables for substituting type variables when
+    /// comparing foralls. Those ids never escape the type equality computations and are used
+    /// solely as rigid type variables: this is why they don't need proper allocation in the
+    /// unification table or to care about those ids clashing with the one generated by the
+    /// typechecker. Generated type constants simply needs to be unique for the duration of the
+    /// type equality computation.
+    var_uid: usize,
+    /// The current gas remaining for variable substitutions. Once it reaches zero and we encounter
+    /// a variable, we abort the computation and return false.
+    gas: u8,
```

Comment:
After testing, indeed u32/u32 and u32/u8 are the same. I think the latter is best, because `gas` ought to stay small. All in all you are right, but in this specific case I don't think 8 bytes for a single parameter passed around inside a casual function like this one really matter (this state is not held very long in memory). I would do the change if it were atomic, but unification variables are already `usize` elsewhere (as they are used to index inside the unification table), so using an u32 here would either need some conversions or to change the type of `TypeWrapper::XXX` variants. I propose to  keep it like this for now